### PR TITLE
feature/TSS-1365-refactor-history-endpoints-v2-part3-v2

### PR DESCRIPTION
**Pull request by osimuka**

### CODE REVIEW

File: api/history/v2/enrichment.py [hunk 1]
```diff
@@ -195,3 +198,211 @@ def enrich(value):
...
+def enrich_public_barrier_status(history: List[Dict]):
+    def is_resolved(status: str):
+        return status == BarrierStatus.RESOLVED_IN_FULL
+
+    def enrich(value):
+        if value and value.get("status"):
+            value["status"] = {
+                "status": str(value["status"]),
+                "status_date": value["status_date"],
+                "is_resolved": is_resolved(str(value["status"])),
```
Comment (chris-pettinga) [inline api/history/v2/enrichment.py:399]: why is `is_resolved` in its own function?

File: api/history/v2/service.py [hunk 2]
```diff
@@ -78,9 +94,32 @@ def enrich_full_history(
...
 
     enriched_history = (
-        barrier_history + programme_fund_history + top_priority_summary_history
+        barrier_history
+        + programme_fund_history
+        + top_priority_summary_history
+        + wto_history
+        + team_member_history
+        + public_barrier_history
+        or [] + public_barrier_notes_history
```
Comment (chris-pettinga) [inline api/history/v2/service.py:121]: I'm not sure I understand the `or` here, let's talk through tomorrow

File: api/collaboration/models.py [hunk 3]
```diff
@@ -50,3 +50,22 @@ def created_user(self):
     @property
     def modified_user(self):
         return self._cleansed_username(self.modified_by)
+
+    @classmethod
+    def get_history(cls, barrier_id, start_date=None):
```
Comment (chris-pettinga) [inline api/collaboration/models.py:55]: The other `get_history` methods do not implement the `start_date` argument, am I missing something?

**Review by chris-pettinga — CHANGES_REQUESTED** [review]
Let's talk through tomorrow after SU

File: api/collaboration/models.py [hunk 4]
```diff
@@ -50,3 +50,29 @@ def created_user(self):
...
+        # due to circlar import, we need to import here
+        from api.history.v2.service import get_model_history
```
Comment (Santino-Trade) [inline api/collaboration/models.py:57]: I came across this too on the Interaction model, but I imported
```python
from api.history.v2 import service
``` 
at the top in the imports and used 
```python
return service.get_model_history(
            qs,
            model="note",
            fields=fields,
            track_first_item=True,
        )
```
in the get_history method. Its a bit neater to keep imports all in their own zone, so i'd say give it a go to see if it works this way for public barriers.

File: api/collaboration/models.py (hunk 3 above)
Comment (Santino-Trade) [reply api/collaboration/models.py:55]: Following it back, it's used to limit the results to everything after the barriers 'reported_on' date. Not sure how that will change your result, but its possible that when the object is created it creates a bunch of empty changes that go from None -> "" or []. I reckon its worth exploring if we can cut it all out.

File: api/collaboration/models.py [hunk 5]
```diff
@@ -50,3 +50,29 @@ def created_user(self):
...
+
+        qs = (
+            cls.history.filter(
+                barrier_id=barrier_id, history_date__gte=start_date
+            ).order_by("id")
+            if start_date
```
Comment (chris-pettinga) [inline api/collaboration/models.py:63]: remove start_date logic and just keep `track_first_time`

File: api/history/manager.py [hunk 6]
```diff
@@ -93,10 +97,38 @@ def get_full_history(cls, barrier, ignore_creation_items=False, use_cache=False)
...
         )
 
+        v2_wto_history = WTOProfile.get_history(
+            barrier_id=barrier.pk, start_date=start_date
+        )
+
+        v2_team_member_history = (
+            TeamMember.get_history(barrier_id=barrier.pk, start_date=start_date)
+            if start_date
+            else TeamMember.get_history(barrier_id=barrier.pk)
```
Comment (chris-pettinga) [inline api/history/manager.py:107]: just keep this line as per above

File: api/history/manager.py [hunk 7]
```diff
@@ -93,10 +97,38 @@ def get_full_history(cls, barrier, ignore_creation_items=False, use_cache=False)
             barrier_id=barrier.pk
         )
 
+        v2_wto_history = WTOProfile.get_history(
+            barrier_id=barrier.pk, start_date=start_date
```
Comment (chris-pettinga) [inline api/history/manager.py:101]: remove start_date from here and make sure `track_first_item` is true

File: api/history/manager.py [hunk 8]
```diff
@@ -93,10 +97,38 @@ def get_full_history(cls, barrier, ignore_creation_items=False, use_cache=False)
...
+            TeamMember.get_history(barrier_id=barrier.pk, start_date=start_date)
+            if start_date
+            else TeamMember.get_history(barrier_id=barrier.pk)
+        )
+
+        v2_public_barrier_history = None
+        v2_public_barrier_notes_history = None
+
+        if barrier.has_public_barrier:
+            if ignore_creation_items:
```
Comment (chris-pettinga) [inline api/history/manager.py:114]: the value of `track_first_item` should be equal to the opposite of the value of `ignore_created_items`

File: api/history/v2/enrichment.py (hunk 1 above)
Comment (chris-pettinga) [reply api/history/v2/enrichment.py:401]:
```suggestion
                "is_resolved": str(value["status"] == BarrierStatus.RESOLVED_IN_FULL,
```

File: api/history/v2/enrichment.py [hunk 9]
```diff
@@ -195,3 +200,240 @@ def enrich(value):
...
+        item["old_value"] = enrich(item["old_value"])
+        item["new_value"] = enrich(item["new_value"])
+
+
+def enrich_team_member_user(history: List[Dict]):
+    def enrich(value, matching_item_dict, is_old):
+        key = "old_value" if is_old else "new_value"
+        if value and isinstance(value, int):
+            user_model = get_user_model()
+            user = user_model.objects.filter(id=value).first()
```
Comment (chris-pettinga) [inline api/history/v2/enrichment.py:418]:
```suggestion
            user = user_model.objects.get(id=value)
```

File: api/interactions/models.py [hunk 10]
```diff
@@ -208,6 +208,31 @@ def created_user(self):
     def modified_user(self):
         return self._cleansed_username(self.modified_by)
 
+    @classmethod
+    def get_history(cls, barrier_id, start_date=None):
```
Comment (chris-pettinga) [inline api/interactions/models.py:212]: can remove start_date

File: api/interactions/models.py [hunk 11]
```diff
@@ -208,6 +208,31 @@ def created_user(self):
...
+
+        fields = (
+            "text",
+            "archived",
+        )
+        return get_model_history(
+            qs,
+            model="public_barrier_note",
+            fields=fields,
+        )
```
Comment (chris-pettinga) [inline api/interactions/models.py:234]: add `track_first_item`

File: api/wto/models.py [hunk 12]
```diff
@@ -94,3 +94,34 @@ class WTOProfile(models.Model):
     case_number = models.CharField(max_length=MAX_LENGTH, blank=True)
 
     history = HistoricalRecords(bases=[WTOProfileHistoricalModel])
+
+    @classmethod
+    def get_history(cls, barrier_id, start_date=None):
```
Comment (chris-pettinga) [inline api/wto/models.py:99]: same as above

**Review by chris-pettinga — CHANGES_REQUESTED** [review]

**Review by Santino-Trade — APPROVED** [review]

**Review by Santino-Trade — APPROVED** [review]

**Review by chris-pettinga — APPROVED** [review]

**Pull request merged**